Source/WebCore/style/StyleBuilderState.cpp

```diff
@@ -332,29 +333,69 @@ double BuilderState::lookupCSSRandomBaseValue(const CSSCalc::RandomCachingKey& k
     return document().lookupCSSRandomBaseValue(key);
 }
 
-unsigned BuilderState::siblingCount() const
+// MARK: - Tree Counting Functions
+
+// "When used on an `element-backed pseudo-element` which is also a real element,
+// the tree counting functions resolve for that real element. For other pseudo
+// elements, they resolve as if they were resolved against the
+// originating element. It follows that for nested pseudo elements the resolution
+// will recursively walk the originating elements until a real element is found.
+//
+// A tree counting function is a `tree-scoped reference` where it references
+// an implicit `tree-scoped name` for the element it resolves against. This is
+// done to not leak tree information to an outer `tree`. A tree counting
+// function that is scoped to an outer `tree` relative to the element it
+// resolves against, will alway resolve to 0."
+//
+//   https://drafts.csswg.org/css-values-5/#tree-counting
+
+unsigned BuilderState::siblingCount()
 {
     // https://drafts.csswg.org/css-values-5/#funcdef-sibling-count
 
     ASSERT(element());
 
-    unsigned count = 1;
-    for (const auto* sibling = ElementTraversal::previousSibling(*element()); sibling; sibling = ElementTraversal::previousSibling(*sibling))
-        ++count;
-    for (const auto* sibling = ElementTraversal::nextSibling(*element()); sibling; sibling = ElementTraversal::nextSibling(*sibling))
-        ++count;
+    // FIXME: This is not exactly what we need. We something that will return nullptr if the parent would have been in an outer shadow scope. In that case, we should return 0;
+    RefPtr parent = element()->parentElementInComposedTree();
+    if (!parent)
+        return 0;
+
+    m_style.setUsesTreeCountingFunctions();
+    parent->setChildrenAffectedByBackwardPositionalRules();
+    parent->setChildrenAffectedByForwardPositionalRules();
+
+    unsigned count = 0;
+    for (auto& node : composedTreeChildren(*parent)) {
```

Comment:
Ditto. Based on the spec text, we probably want to use regular children, not children in the composed tree.